### ISSUE TITLE
Connection pool with multiple hosts

### ISSUE THREAD

**Issue by neumino** [neumino/rethinkdbdash#71]
I'm looking for feedback on the connection pool with multiple hosts.

__Context__:
Having a connection pool connecting to multiple hosts is something pretty nice to have:
- It's easier to think with a concrete example. Suppose you have a cluster of 3 machines `M1`, `M2` and `M3`. Your client `C` is connected to `M1`.
Suppose that the connection `M1`-`C` goes down, but not the connection `M<x>`-`M<y>`
In this case, you could connect `C` to `M2` and everything will work.
- RethinkDB is implementing raft, and auto failover will be shipped in the next few months (not sure exactly when). At that time a machine could just disappear, and the cluster will still work fine. It would be bad if the point of failure becomes the driver.

So having the pool connected to multiple server should help a lot.

__Implementation__:
The implementation I'm working on has the following features:
- There is basically a "pool master", and `n` pools (if you connect to `n` different servers)/
- You can connect to as many hosts as you want with whatever options you want (`buffer`/`size`)
- If a server goes down, or if the connection client-server is down, its pool will become empty (all the connections will be closed). At that time, the pool is considered not healthy, and all the quotas of connections it has, is spread across the other servers.
So if you have these settings:
```
M1: buffer=200, max=1000
M2: buffer=80, max=300
M3: buffer=150, max=900
```
And if `M1` goes down, then the settings become:
```
M2: buffer=`80, max=800
M3: buffer=250, max=1400
```

The idea behind this behavior is that the driver should keep the same number of connections available/max connections across all the servers to be able to sustain the same workload.
- If a server comes back (the pool will try to recreate connection with an exponential back off like it does now), the pool will take back its quotas and things come back to normal.

- Queries are still executed with `run()`. You do not have an explicit access to the pool and you can't specify on which server the query should be sent.
- Queries are sent to different servers in a robin hood fashion. Before sending a query to a server, we check that the pool has an available connection. If it doesn't, we'll try to check another server. If no connection is available, we wait for a connection to be freed except if all the pools have errors, in which case we'll reject the query.

__Limitations__:
- You cannot explicitly select a pool - one reasonable use case is to route query directly to the appropriate master but that seems stretched.
- You cannot update the servers to which the driver is connecting to on the fly. Should we pull the servers from `r.db('rethinkdb').table('server_status')`? And let the driver create the pools?

Any thoughts on that? 

Site note: As part of this work, I've also fixed #50 by adding reqlite in the tests.

**Comment by hugufc** [comment]
I couldn't understand something:

Considering there is a master pool, so M1, M2 and M3 are under one master pool localized in other server? Or one of them is the master pool (How they know about each other)?

Let me know if this question make senses.

**Comment by neumino** [comment]
Sorry for the confusion.

What I've called master pool is a class in the driver. It's just the code responsible to decide which connection pool to use.

I also forgot the syntax. It would look like this for now:
```
var r = require('rethinkdbdash')({
    hosts: [
        {host: "server1_ip", port: 28015},
        {host: "server2_ip", port: 28015},
        {host: "server3_ip", port: 28015},
    ]
})
```
Maybe `servers` is better than `hosts`?

**Comment by hugufc** [comment]
Thanks for the explanation @neumino,

BTW, `servers` looks good to me, considering that host is a prop. Like [hapijs](www.hapijs.com).

**Comment by neumino** [comment]
So there's a discovery mode (by default true) where the pool will pull data from `rethinkdb_status` and connect to all hosts in the cluster.
There's a bit of work left:
- Properly handle localhost as any machine could be localhost
- Handle proxys and similar things (admin?)

**Comment by neumino** [comment]
Fixed in 2.0.0
